dnsmasq: fix resolv.conf for round-robin DNS configuration

feltai opened the pull request:
Adds the ability to configure what types of addresses (IPv4/IPv6) are used in resolv.conf to query the local nameserver.
This is achieved by using `add_local_ipv4_nameserver` and `add_local_ipv6_nameserver` options in dhcp config file, dnsmasq configuration section.

Default behavior change:
Now only one localhost IPv4 address is used in resolv.conf to query the local nameserver instead of IPv4 and IPv6 localhost addresses (`127.0.0.1` and `::1`).

Fixes resolv.conf to not query the same nameserver twice. Querying the same nameserver twice introduces issues
with answers for records that have multiple IP values. Those types of records are used for high availability by utilizing round-robin DNS for client side load balancing.

Additional information on how musl libc resolver works:

Name Resolver/DNS section:
https://wiki.musl-libc.org/functional-differences-from-glibc.html

Comments from the lead developer of musl libc Rich Felker:
https://www.openwall.com/lists/musl/2022/07/11/1

Fixes: https://github.com/openwrt/openwrt/issues/10278

feltai commented:
What is missing to get this merged?

Ansuel commented:
@feltai
Hi,
- This lacks of Sob tag
- If this fix an issue you should use the `Fixes:` tag 
- You should add a commit description to the commit to describe what it does and why (aside from fixing the linked issue)

:D

feltai commented:
@Ansuel Can You take a look at my updated commit?

Ansuel commented:
@feltai Much better but I think a better naming would be `dnsmasq: fix resolv.conf for round-robin DNS configuration`
Also you should use` Fixes: https://github.com/openwrt/openwrt/issues/10278` to actually link the commit to the issue. (github thing) (also update the pr title and description to reflect the new commit)

Now to the real question:

I wonder if removing ::1 would cause some problems tho.
I like the idea of making this configurable.

Should we keep the default behavior and document this in the wiki? (aka not disabling ::1 by default?)
Or should we add some logic and auto-chose the correct value based on the choice? My concern is with some crazy configuration where ipv4 is disabled.

@jow- @neheb wonder if you want to give some feedback about this?

feltai commented:
Updated commit and PR text.

neheb commented on package/network/services/dnsmasq/files/dnsmasq.init:
```diff
@@ -1143,8 +1147,12 @@ dnsmasq_start()
 		}
-		DNS_SERVERS="$DNS_SERVERS 127.0.0.1"
-		[ -e /proc/sys/net/ipv6 ] && DNS_SERVERS="$DNS_SERVERS ::1"
+		[ $ADD_LOCAL_IPV4_NAMESERVER -eq 1 ] && {
```
Needs quoting here. shellcheck will definitely warn about this.

BKPepe commented:
Shouldn't this go to the master branch, first?

feltai commented:
@neheb I checked the init script with shellcheck but there were no errors or warnings for those lines of code.

neheb commented:
uhh:

```
[ -e /proc/sys/net/ipv6 ] && [ $ADD_LOCAL_IPV6_NAMESERVER -eq 1 ]
                               ^------------------------^ SC2086 (info): Double quote to prevent globbing and word splitting.
      
```

feltai commented:
@neheb quoted the variables. Should I change the PR target branch to `master` instead of `openwrt-22.03`?

neheb commented:
Yes

feltai commented:
Done

feltai committed:
dnsmasq: fix resolv.conf for round-robin DNS configuration

Adds the ability to configure what types of addresses (IPv4/IPv6)
are used in resolv.conf to query the local nameserver.
This is achieved by using 'add_local_ipv4_nameserver' and
'add_local_ipv6_nameserver' options in dhcp config file,
dnsmasq configuration section.

Default behavior change:
Now only one localhost IPv4 address is used in resolv.conf
to query the local nameserver instead of IPv4 and IPv6
localhost addresses (127.0.0.1 and ::1).

Fixes resolv.conf to not query the same nameserver twice.
Querying the same nameserver twice introduces issues
with answers for records that have multiple IP values.
Those types of records are used for high availability
by utilizing round-robin DNS for client side load balancing.

Additional information on how musl libc resolver works:

Name Resolver/DNS section:
https://wiki.musl-libc.org/functional-differences-from-glibc.html

Comments from the lead developer of musl libc Rich Felker:
https://www.openwall.com/lists/musl/2022/07/11/1

Fixes: https://github.com/openwrt/openwrt/issues/10278

Signed-off-by: Felix Tailor <felix.tailor@pm.me>

feltai commented:
@Ansuel @neheb Is this PR ready for the workflow approval?

Neustradamus commented:
@feltai: Any progress on your PR?
